security/acme-client/src/opnsense/scripts/OPNsense/AcmeClient/certhelper.php

```diff
@@ -1203,20 +1257,62 @@ function log_cert_acme_status($certObj, $modelObj, $statusCode)
 function dump_postponed_updates()
 {
     global $postponed_updates;
+
+    $status_descr = [
+        100 => 'unknown',
+        200 => 'OK',
+        250 => 'cert revoked',
+        300 => 'configuration error',
+        400 => 'validation failed',
+        500 => 'internal error',
+    ];
+
     $modelObj = new OPNsense\AcmeClient\AcmeClient;
 
     foreach ($postponed_updates as $pupdate) {
-        $node = $modelObj->getNodeByReference('certificates.certificate.' . $pupdate['uuid']);
+        $_statusCode = $pupdate['statusCode'];
+        $_uuid = $pupdate['uuid'];
+        $node = $modelObj->getNodeByReference('certificates.certificate.'.$_uuid);
         if ($node != null) {
-            $node->statusCode = $pupdate['statusCode'];
+            log_error("AcmeClient: storing status '" . $status_descr[$_statusCode] . "' for cert " . (string)$node->name);
+            $node->statusCode = $_statusCode;
             $node->statusLastUpdate = $pupdate['statusLastUpdate'];
             // serialize to config and save
             $modelObj->serializeToConfig();
             Config::getInstance()->save();
+            Config::getInstance()->forceReload();
+        } else {
+            log_error(sprintf("AcmeClient: failed to store status '%s' for cert %s: node not found",$status_descr[$_statusCode],$_uuid));
         }
     }
 }
 
+function run_shell_command($proc_cmd, $proc_env = array())
```

Comment:
@fraenki it's better to rewrite this to configd usage, if sometime in the future the webui won't have root access, this might fail (we try to avoid shell exec in our new code base as much as possible).

Reply:
@AdSchellevis, thanks for the review. I like your suggestion, but there are two things to consider:

1. No command in our Let's Encrypt plugin requires root privileges. They are only run as root because this is how it works today.
1. The Google Cloud CLI requires that certain environment variables are set [1]. This is not supported by configd, AFAIK.

The latter probably prevents using configd for this purpose.

[1] https://github.com/opnsense/plugins/pull/1318/files/4132244bb050495a1f72114661f9ad7b1fe1a2a1#diff-6357d5c08a4b4115e3aea0d88e069eb1R689

Reply:
@fraenki our usual approach for these things is to detach both processes, so you have a program (script) that performs an action (like your gcloud setup), which uses some configuration data (usually in the form of a config/ini file or parameters if limited) and is self contained. Configd calls the action.

The main advantage of this approach is that it eases testing a lot (you can always write a unit test if needed afterwards) and ensures your code base is easier to maintain.

In this case I would probably create a configuration template to flush the settings into an ini (or json) file, which can then be picked up from your gcloud init script. Step by step your program would do the following:

1. [controller] refresh template, using configd (which writes the configuration gcloud needs)
2. [controller] exec configd action "dns_gcloud"
3. [configd] executes attached script, reports status
4. [script] read settings (from config flushed in 1)
5. [script] do actions, probably these steps: https://github.com/opnsense/plugins/blob/4132244bb050495a1f72114661f9ad7b1fe1a2a1/security/acme-client/src/opnsense/scripts/OPNsense/AcmeClient/certhelper.php#L684-L696
6. [controller] report status

The script can incorporate its own logging, can be supplied with a sample configuration file for testing purposes if needed, and could be started separately while developing/testing logic.

I'm not saying that the current code doesn't work, just noticing that maintainability could be improved and trying to prevent future disappointments (breakage after changes in core at some point in time).

Reply:
> our usual approach for these things is to detach both processes

Understood, and I fully agree with your suggestion. I just want to point out that the current state of `certhelper.php` has no future anyway (at least in my opinion, it's a piece of spaghetti code that should be replaced/refactored rather sooner than later). I don't want to work on a half-baked approach that would only be implemented for _1_ out of the 47 supported DNS services.

I want to rethink the approach and rewrite `certhelper.php`, but I ask you for some more time. I don't want to start this lengthy process in the middle of implementing one of the oldest feature requests and fixing possibly the most-reported bug.

@AdSchellevis, if you agree I'd like to open a new issue for the rewrite, add your suggestions as a starting point, and merge the current state for now. Your points are all valid, but I need time to work on this.

Reply:
@fraenki no problem at all, I might have misread some portions as well, sometimes its best to refactor once it's more clear what the code actually should do. We all trash code from time to time :) shall I merge this now?